packages/headless/src/controllers/core/folded-result-list/headless-core-folded-result-list.ts

```diff
@@ -0,0 +1,274 @@
+import {AsyncThunkAction} from '@reduxjs/toolkit';
+import {
+  ClickAction,
+  CustomAction,
+} from '../../../features/analytics/analytics-utils';
+import {configurationReducer as configuration} from '../../../features/configuration/configuration-slice';
+import {
+  foldingOptionsSchemaDefinition,
+  registerFolding,
+} from '../../../features/folding/folding-actions';
+import {foldingReducer as folding} from '../../../features/folding/folding-slice';
+import {
+  CollectionId,
+  FoldedCollection,
+  FoldedResult,
+} from '../../../features/folding/folding-state';
+import {queryReducer as query} from '../../../features/query/query-slice';
+import {searchReducer as search} from '../../../features/search/search-slice';
+import {CoreEngine, Result} from '../../../recommendation.index';
+import {
+  ConfigurationSection,
+  FoldingSection,
+  SearchSection,
+} from '../../../state/state-sections';
+import {loadReducerError} from '../../../utils/errors';
+import {validateOptions} from '../../../utils/validate-payload';
+import {Controller} from '../../controller/headless-controller';
+import {
+  ResultListOptions,
+  buildCoreResultList,
+} from '../result-list/headless-core-result-list';
+import {SearchStatusState} from '../status/headless-core-status';
+
+export type {FoldedCollection, FoldedResult};
+
+const optionsSchema = new Schema<Required<FoldingOptions>>(
+  foldingOptionsSchemaDefinition
+);
+
+export interface FoldingOptions {
+  /**
+   * The name of the field on which to do the folding. The folded result list component will use the values of this field to resolve the collections of result items.
+   *
+   * @defaultValue `foldingcollection`
+   */
+  collectionField?: string;
+  /**
+   * The name of the field that determines whether a certain result is a top result containing other child results within a collection.
+   *
+   * @defaultValue `foldingparent`
+   */
+  parentField?: string;
+  /**
+   * The name of the field that uniquely identifies a result within a collection.
+   *
+   * @defaultValue `foldingchild`
+   */
+  childField?: string;
+  /**
+   * The number of child results to fold under the root collection element, before expansion.
+   *
+   * @defaultValue `2`
+   */
+  numberOfFoldedResults?: number;
+}
+
+export interface FoldedResultListOptions extends ResultListOptions {
+  folding?: FoldingOptions;
+}
+
+export interface CoreFoldedResultListProps {
+  /**
+   * The options for the `FoldedResultList` controller.
+   * */
+  options?: FoldedResultListOptions;
+  /**
+   * The action creator to build the `loadCollection` action.
+   */
+  loadCollectionActionCreator: (
+    collectionId: CollectionId
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ) => AsyncThunkAction<any, CollectionId, any>;
+}
+
+/**
+ * The `CoreFoldedResultList` headless controller re-organizes results into hierarchical collections (a.k.a. threads).
+ */
+export interface CoreFoldedResultList extends Controller {
```

Comment:
good point, I will refactor these, only thing is the props which I will have to export from the files because of the actionCreator pattern. It does clean up the files nicely though thanks!